apiserver/root.go

```diff
@@ -183,23 +184,58 @@ func rpcRoot(srv *Server, root *apiHandler, authTag names.Tag) (rpc.Root, error)
 		root.resources,
 		root,
 	)
+	return maybeRestrictRoot(srv, apiRoot, root.state.IsController(), authTag)
+}
+
+func maybeRestrictRoot(
+	srv *Server,
+	apiRoot rpc.Root,
+	isControllerModel bool,
+	authTag names.Tag,
+) (rpc.Root, error) {
+
+	if isControllerModel && authTag == srv.tag {
+		// The local controller machine agent is
+		// always allowed to connect to itself.
+		//
+		// TODO(axw) allow all controller machine
+		// agents, using the source address?
+		// See lp:1733259.
+		return apiRoot, nil
```

Comment:
We have the tags, we could check the jobs, and don't really need to know from the IP address, do we?
Note that containers might appear to be coming from a similar IP address depending on nat, etc.

Also any unit agents deployed to controller machines would also have the same IP address.

Reply:
Good point on other agents sharing the address. I said that because this code is called prior to doing any rate limiting or authentication.

We can most likely shuffle things around a bit to make it work, though. I'll remove the comment about source address.